### PR TITLE
Rewrite exercise 2

### LINKED ISSUE #46: Rewrite exercise 2
- rewrite to TypeScript
- add end-to-end tests

### PR DESCRIPTION
**Pull request by dinassh**
## This PR rewrites exercise 2:
- [x] change language to TypeScript
- [x] add smoke tests
- [x] add solutions to additional exercises
- [x] rename `UsersList` to `UserList`, `UsersTypes` to `UserTypes`, also file names
- [x] change `Array<Type>` to `Type[]`
- [x] remove default exports
- [x] ~~possibly remove `React.FC` type, see https://github.com/salsita/react-training/pull/44#discussion_r460868075~~
- [x] ~~add readonly types, based on the https://github.com/salsita/react-training/pull/44#discussion_r460959577 discussion~~
- [x] fix tsconfig.json to use `extends` syntax 
- [x] update `.github/workflows/main.yml` with changes introduced in https://github.com/salsita/react-training/pull/44
- [x] add type explanation to tutorial

### CODE REVIEW

File: exercises/02-react-stateless/src/modules/root/components/header.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+
+interface HeaderProps {
+  title: string
+}
+
+export const Header: React.FC<HeaderProps> = ({ title }) => {
+  return <h1>{title}</h1>
+}
```
Comment (branoholy) [inline exercises/02-react-stateless/src/modules/root/components/header.tsx]: Not necessary to change this but simple functions like this one don't need `return`.

```suggestion
export const Header: React.FC<HeaderProps> = ({ title }) => <h1>{title}</h1>
```

File: exercises/02-react-stateless/src/modules/users/user-types.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  lastName: string
+}
+
+export interface User extends UserData {
+  id: number
+}
+
+export interface AddUserFunc {
+  (user: UserData): void
+}
```
Comment (branoholy) [inline exercises/02-react-stateless/src/modules/users/user-types.ts]: I prefer `type` for functions

```suggestion
export type AddUserFunc = (user: UserData) => void
```

**Review by branoholy — APPROVED** [review]

File: exercises/02-react-stateless/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+
+import { Root } from 'modules/root/components/root'
+import { User, AddUserFunc } from 'modules/users/user-types'
```
Comment (novellizator) [inline exercises/02-react-stateless/src/index.tsx]: Let's not use hungarian notation (AddUser should be enough)

Comment (dinassh) [reply exercises/02-react-stateless/src/index.tsx]: Renamed, see 9734160

**Review by novellizator — APPROVED** [review]
LGTM, merge please!

**Pull request merged**